File: src/DOM.ts
import * as EventEmitter from 'events'
import * as WebKit from 'WebKit2'

export default class DOM extends EventEmitter {
  webView: WebKit.WebView

  constructor(webView) {
    super()
    this.webView = webView
  }

  /* High-level methods */

  click(selector) {
    this.exec('click', selector)
  }

  triggerAction(action) {
    this.click(`[jsaction="${action}"]`)
  }

  setValue(selector, value) {
    this.exec('setValue', selector, value)
  }

  getValue(selector) {
    // TODO: Implement returning values directly, using async, or a callback
    this.exec('getValue', selector)
  }

  /* Helper methods */

  exec(method, ...params) {
    const paramsString = params
      .map(param => JSON.stringify(param))
      .join(', ')
    const code = `domHelper.${method}(${paramsString})`

    this.webView.runJavaScript(code, null, (webView, result) => {
      console.log(this.webView.runJavaScriptFinish(result))
    })
  }
}

File: src/dom.rs
use itertools::join;
use webkit2gtk::{WebView, WebViewExtManual};

pub fn click(webview: &WebView, selector: &str) {
    exec(webview, "click", &[selector]);
}

pub fn trigger_action(webview: &WebView, action: &str) {
    click(webview, &format!("[jsaction=\"{}\"]", action));
}

pub fn set_value(webview: &WebView, selector: &str, value: &str) {
    exec(webview, "setValue", &[selector, value]);
}

// pub fn get_value(webview: &WebView, selector: &str) {
//     exec(webview, "getValue", selector);
// }

fn exec(webview: &WebView, method: &str, params: &[&str]) {
    let params = join(params.iter().map(|param| format!("'{}'", param)), ", ");
    let code = format!("domHelper.{}({})", method, params);

    webview.run_javascript_with_callback(&code, |result| match result {
        Ok(_) => {}
        Err(error) => println!("WARNING: {}", error),
    });
}
